File: PakingLotDemo/src/strategy/parking/ParkingStrategy.java
package strategy.parking;

import model.ParkingFloor;
import model.ParkingSpot;
import model.Vehicle;

import java.util.List;

/**
 * Strategy for selecting a parking spot for a vehicle (e.g. nearest first, best fit).
 */
public interface ParkingStrategy {

    /**
     * Finds an available spot that fits the vehicle, or null if none.
     *
     * @param floors  the list of floors to search
     * @param vehicle the vehicle to park
     * @return an available spot that fits, or null if parking is full
     */
    ParkingSpot findSpot(List<ParkingFloor> floors, Vehicle vehicle);
}

File: problems/python/ParkingLotSystem/src/strategy/parking/parking_strategy.py
"""Strategy for selecting a parking spot."""
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import Optional
from model.parking_floor import ParkingFloor
from model.parking_spot import ParkingSpot
from model.vehicle import Vehicle


class ParkingStrategy(ABC):
    @abstractmethod
    def find_spot(
        self, floors: list[ParkingFloor], vehicle: Vehicle
    ) -> Optional[ParkingSpot]:
        pass
